src/shamir.rs

```diff
@@ -0,0 +1,398 @@
+//!
+//! This module contains the implementation of [`Shamir secret sharing`] using finite fields.
+//! It uses FF implementations provided by [`ff`] crate and is mostly useful when working with
+//! prime fields as they support arithmetic addition and multiplication of shares that is used
+//! to implement secure multiplication protocol.
+//!
+//! [`Shamir secret sharing`](https://en.wikipedia.org/wiki/Shamir%27s_Secret_Sharing)
+//! [`ff`](https://crates.io/crates/ff)
+//!
+use ff::{Field, PrimeField};
+use rand_core::RngCore;
+use std::iter::repeat_with;
+use std::num::NonZeroU8;
+use thiserror::Error;
+
+/// Shamir secret sharing
+pub struct SecretSharing {
+    /// Threshold
+    k: u8,
+
+    /// Number of shares produced by the `share` method
+    n: u8,
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Wrong or insecure secret sharing scheme. (expected {k} >= {n} > 1)")]
+    BadSharingScheme { k: u8, n: u8 },
+    #[error("Not enough shares to reconstruct secret: need at least {actual}, got {required}")]
+    NotEnoughShares { required: u8, actual: u8 },
+    #[error("Prime field element inversion failed: {v:?}")]
+    InvertError { v: Box<[u8]> },
+}
+
+/// Represents a single share: (x, f(x)) point
+#[derive(Clone)]
+pub struct Share<F> {
+    x: u8,
```

Comment:
Each helper will only ever have just one value of "x" for all of its shares. It seems like a waste of memory to copy it around so much. Maybe you can just store this value in a single place and reference it when you need it.

Reply:
it is possible, but more expensive as we would need to keep an 8 byte pointer to 1 byte value. Also, if we want helper Y to only work with shares "x = Y" then we would need to validate that all shares it receives from outside have "x" set to Y and reject such inputs as soon as it discovers that, so `Share` object carrying `x` value makes it possible

Reply:
I still think it's un-necessary. Each helper should just know what its x-coordinate it. It doesn't need a pointer to it in every secret share.

When shares are received from other helpers for the purpose of reconstructing the secret - it just needs to know which batch of shares came from which helper.

But this is an optimization we can always add later. This version has more opportunities for validation which can assist with debugging.

Reply:
it is done now